libctru/source/allocator/mappable.cpp

```diff
@@ -0,0 +1,74 @@
+extern "C"
+{
+	#include <3ds/types.h>
+	#include <3ds/mappable.h>
+	#include <3ds/util/rbtree.h>
+}
+
+#include "mem_pool.h"
+#include "addrmap.h"
+
+extern u32 __mappable_heap, __mappable_heap_size;
+
+static MemPool sMappablePool;
+
+static bool mappableInit()
+{
+	auto blk = MemBlock::Create((u8*)__mappable_heap, __mappable_heap_size);
+	if (blk)
+	{
+		sMappablePool.AddBlock(blk);
+		rbtree_init(&sAddrMap, addrMapNodeComparator);
+		return true;
+	}
+	return false;
+}
+
+void* mappableAlloc(size_t size)
+{
+	size_t alignment = 0x1000; // hardcoded to page alignment
+	// Convert alignment to shift amount
+	int shift;
+	for (shift = 4; shift < 32; shift ++)
+	{
+		if ((1U<<shift) == alignment)
+			break;
+	}
+	if (shift == 32) // Invalid alignment
+		return nullptr;
+
+	// Initialize the pool if it is not ready
+	if (!sMappablePool.Ready() && !mappableInit())
+		return nullptr;
+
+	// Allocate the chunk
+	MemChunk chunk;
+	if (!sMappablePool.Allocate(chunk, size, shift))
```

Comment:
This could have been simplified to `if (!sMappablePool.Allocate(chunk, size, 12)) // Force page alignment`, and the above section that calculates the shift removed.